packages/eslint-plugin-stories/src/rules/csf-v3-types.ts

```diff
@@ -0,0 +1,160 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/experimental-utils';
+import isStories from '../utils/isStories';
+
+/**
+ * Enforce specifying an explicit type for story objects where necessary for `composeStories` and
+ * Storybook to correctly infer args.
+ *
+ * Currently only detects 2 cases:
+ *
+ * 1. Empty objects
+ *    ```
+ *    const Primary = {} // <- violation
+ *    export const Secondary: StoryObj<Args> = {} // <- ok
+ *    ```
+ *
+ * 2. Objects with a render function that takes args
+ *    ```
+ *    export const Primary = { // <- violation
+ *      render: (args) => <Button {...args} />,
+ *    }
+ *    ```
+ *
+ * These should cover most of the times we'll need explicit types. Other cases do exist, such as
+ * objects that spread in another but don't define their own render function
+ *
+ *   ```
+ *   export const Secondary = { // <- Might be a violation, but we don't flag it
+ *     ...Primary,
+ *   }
+ *   ```
+ *
+ * But these other cases aren't worth jumping through hoops to find.
+ */
+const rule: TSESLint.RuleModule<'specifyType'> = {
+  meta: {
+    messages: {
+      specifyType:
+        'Specify a type of `StoryObj<Args>` for object stories, so that `composeStories` and Storybook can infer the correct args',
+    },
```

Comment:
Using `TSESLint.RuleModule` is necessary to get the right TS types for the AST nodes. Well, that or some type casting.

What's weird about `TSESLint.RuleModule` is it doesn't let you do

```js
context.report({
  node,
  message: 'some message',
});
```

Instead it forces you (via the type system) to specify the message in `meta` and pass in the message id.

At runtime it works fine to pass in `message`, but I wonder if the folks at `@typescript-eslint` really wanted to enforce usage of `messageId` for some reason.